src/rate_limiter/src/lib.rs

```diff
@@ -357,23 +374,40 @@ impl RateLimiter {
             TokenType::Ops => self.ops.as_mut(),
         };
         // Try to consume from the token bucket.
-        let success = match token_bucket {
-            Some(bucket) => bucket.reduce(tokens),
+        match token_bucket {
+            Some(bucket) => {
+                let refill_time = bucket.refill_time_ms();
+                match bucket.reduce(tokens) {
+                    // When we report budget is over, there will be no further calls here,
+                    // register a timer to replenish the bucket and resume processing;
+                    // make sure there is only one running timer for this limiter.
+                    BucketReduction::Failure => {
+                        self.activate_timer(TIMER_REFILL_STATE);
+                        false
```

Comment:
Actually, this new code changes the relation between timer state (active/inactive) and the return value of the main `consume()` method (success/failure).

With the old code there it was guaranteed that timer would only be active/activated when a `consume()` operation fails. This version of the code activates the timer on `OverConsumption` **while** allowing (returning `true`) on that `consume()` call.

As such, the `activate_timer()` should set default state in case of `BucketReduction::Failure` only if `timer.active == false`.

You can reintroduce the timer state check in `activate_timer()` or wrap it here locally where the reason/need for it is localized:
```suggestion
                        if !self.timer_active {
                            self.activate_timer(TIMER_REFILL_STATE);
                        }
                        false
```

Reply:
Added this as well as a check at the beginning of `consume()` to not let the rate limiter consume any tokens while the timer is active.